base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,26 +20,289 @@
+                int cy = GetSystemMetrics(SM_CYBORDER) * 4;
+                int x = rcItem.left + (rcWnd.left - rcParent.left) - 2;
+                int y = rcItem.top + (rcWnd.top - rcParent.top) - rcHdr.bottom - cy;
+
+                SendMessage(hCombo, CB_SETCURSEL, iItem.lParam, 0);
+                ::SetWindowPos(hCombo, NULL, x, y, 0, 0, SWP_NOSIZE | SWP_SHOWWINDOW);
+                return FALSE;
+            }
+        }
+    }
+    return TRUE;
+}
+
+BOOL TrayIconView::OnCommand(HWND hwnd, WPARAM wParam, LPARAM lParam)
+{
+    switch(LOWORD(wParam))
+    {
+        case IDC_NOTIFICATION_BEHAVIOUR:
+            if (HIWORD(wParam) == CBN_SELCHANGE)
+            {
+                WCHAR szText[128];
+                HWND hComBox = (HWND)lParam;
+                LVITEMW iItem;
+
+                INT idx = SendMessageW(hComBox, CB_GETCURSEL, 0, 0);
+                SendMessageW(hComBox, CB_GETLBTEXT, (WPARAM)idx, (LPARAM)szText);
+
+                INT lvItem = GetNextItem(-1, LVNI_SELECTED);
+
+                iItem.mask = LVIF_TEXT;
+                iItem.cchTextMax = lstrlenW(szText);
+                iItem.pszText = szText;
+                iItem.iItem = lvItem;
+                iItem.iSubItem = 1;
+                ListView_SetItem(hListView, &iItem);
+
+                iItem.mask = LVIF_PARAM;
+                iItem.iItem = lvItem;
+                iItem.iSubItem = 0;
+                iItem.lParam = idx;
+                ListView_SetItem(hListView, &iItem);
+
+                ::ShowWindow(hComBox, SW_HIDE);
+                UpdateIcon(lvItem, idx);
+            }
+            return FALSE;
+
+        case IDOK:
+        case IDCANCEL:
+            SendMessage(hwnd, WM_CLOSE, 0, 0);
+            break;
+    
+        default:
+            return FALSE;
+    }
+    return TRUE;
+}
+
+BOOL TrayIconView::AddColumn(INT Index, INT Width, DWORD strId, INT fmt)
+{
+    CStringW strLabel(MAKEINTRESOURCEW(strId));
+    LVCOLUMNW Column;
+    ZeroMemory(&Column, sizeof(Column));
+
+    Column.mask =  LVCF_WIDTH | LVCF_SUBITEM | LVCF_TEXT | LVCF_FMT;
+    Column.iSubItem = Index;
+    Column.cx = Width;
+    Column.pszText = (LPWSTR)strLabel.GetString();
+    Column.cchTextMax = strLabel.GetLength();
+    Column.fmt = fmt;
+
+    return (InsertColumn(Index, &Column) == -1) ? FALSE : TRUE;
+}
+
+BOOL TrayIconView::InsertItem(LPWSTR data, INT iSubItem, INT id)
+{
+    LVITEMW iItem;
+
+    iItem.mask = LVIF_TEXT | LVIF_IMAGE;
+    iItem.pszText = data;
+    iItem.cchTextMax = lstrlenW(data);
+    iItem.iSubItem = iSubItem;
+    iItem.iItem = id;
+    iItem.iImage = id;
+
+    if (iSubItem)
+        return (ListView_SetItem(m_hWnd, &iItem) != -1);
+    else
+        return (ListView_InsertItem(m_hWnd, &iItem) != -1);
+}
+
+VOID TrayIconView::UpdateIcon(USHORT iItem, USHORT Preference)
+{
+    HWND m_hwndPager = FindWindowHandleByPath(L"Shell_TrayWnd/TrayNotifyWnd/SysPager");
+    NOTIFYICONDATAW *pIcon = &m_TrayIconData[iItem];
```

Comment:
Be careful with this, when resizing `m_TrayIconData`, the address of all items contained in it can change!